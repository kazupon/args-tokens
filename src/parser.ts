/**
 * Entry point of argument parser.
 * @module
 */

/**
 * forked from `nodejs/node` (`pkgjs/parseargs`)
 * repository url: https://github.com/nodejs/node (https://github.com/pkgjs/parseargs)
 * code url: https://github.com/nodejs/node/blob/main/lib/internal/util/parse_args/parse_args.js
 *
 * @author kazuya kawaguchi (a.k.a. kazupon)
 * @license MIT
 */

/**
 * Argument token Kind.
 * - `option`: option token, support short option (e.g. `-x`) and long option (e.g. `--foo`)
 * - `option-terminator`: option terminator (`--`) token, see guideline 10 in https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html
 * - `positional`: positional token
 */
type ArgTokenKind = 'option' | 'option-terminator' | 'positional'

/**
 * Argument token.
 */
export interface ArgToken {
  /**
   * Argument token kind.
   */
  kind: ArgTokenKind
  /**
   * Argument token index, e.g `--foo bar` => `--foo` index is 0, `bar` index is 1.
   */
  index: number
  /**
   * Option name, e.g. `--foo` => `foo`, `-x` => `x`.
   */
  name?: string
  /**
   * Raw option name, e.g. `--foo` => `--foo`, `-x` => `-x`.
   */
  rawName?: string
  /**
   * Option value, e.g. `--foo=bar` => `bar`, `-x=bar` => `bar`.
   * If the `allowCompatible` option is `true`, short option value will be same as Node.js `parseArgs` behavior.
   */
  value?: string
  /**
   * Inline value, e.g. `--foo=bar` => `true`, `-x=bar` => `true`.
   */
  inlineValue?: boolean
}

const HYPHEN_CHAR = '-'
const HYPHEN_CODE = HYPHEN_CHAR.codePointAt(0)!
const EQUAL_CHAR = '='
const EQUAL_CODE = EQUAL_CHAR.codePointAt(0)!
const TERMINATOR = '--'
const SHORT_OPTION_PREFIX = HYPHEN_CHAR
const LONG_OPTION_PREFIX = '--'

/**
 * Parser Options.
 */
export interface ParserOptions {
  /**
   * [Node.js parseArgs](https://nodejs.org/api/util.html#parseargs-tokens) tokens compatible mode.
   * @default false
   */
  allowCompatible?: boolean
}

/**
 * Parse command line arguments.
 * @example
 * ```js
 * import { parseArgs } from 'args-tokens' // for Node.js and Bun
 * // import { parseArgs } from 'jsr:@kazupon/args-tokens' // for Deno
 *
 * const tokens = parseArgs(['--foo', 'bar', '-x', '--bar=baz'])
 * // do something with using tokens
 * // ...
 * console.log('tokens:', tokens)
 * ```
 * @param args command line arguments
 * @param options parse options
 * @returns Argument tokens.
 */
export function parseArgs(args: string[], options: ParserOptions = {}): ArgToken[] {
  const { allowCompatible = false } = options

  const tokens: ArgToken[] = []
  const remainings = [...args]
  let index = -1
  let groupCount = 0
  let hasShortValueSeparator = false

  while (remainings.length > 0) {
    const arg = remainings.shift()

    if (arg == undefined) {
      break
    }

    const nextArg = remainings[0]
    if (groupCount > 0) {
      groupCount--
    } else {
      index++
    }

    // check if `arg` is an options terminator.
    // guideline 10 in https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html
    if (arg === TERMINATOR) {
      tokens.push({
        kind: 'option-terminator',
        index
      })
      const mapped = remainings.map(arg => {
        return { kind: 'positional', index: ++index, value: arg }
      }) satisfies ArgToken[]
      tokens.push(...mapped)
      break
    }

    if (isShortOption(arg)) {
      const shortOption = arg.charAt(1)
      let value: string | undefined
      let inlineValue: boolean | undefined
      if (groupCount) {
        // e.g. `-abc`
        tokens.push({
          kind: 'option',
          name: shortOption,
          rawName: arg,
          index,
          value,
          inlineValue
        })
        if (groupCount === 1 && hasOptionValue(nextArg)) {
          value = remainings.shift()
          if (hasShortValueSeparator) {
            inlineValue = true
            hasShortValueSeparator = false
          }
          tokens.push({
            kind: 'option',
            index,
            value,
            inlineValue
          })
        }
      } else {
        // e.g. `-a`
        tokens.push({
          kind: 'option',
          name: shortOption,
          rawName: arg,
          index,
          value,
          inlineValue
        })
      }

      // eslint-disable-next-line unicorn/no-null
      if (value != null) {
        ++index
      }
      continue
    }

    if (isShortOptionGroup(arg)) {
      // expend short option group (e.g. `-abc` => `-a -b -c`, `-f=bar` => `-f bar`)
      const expanded = []
      let shortValue = ''
      for (let i = 1; i < arg.length; i++) {
        const shortableOption = arg.charAt(i)
        if (hasShortValueSeparator) {
          shortValue += shortableOption
        } else {
          if (!allowCompatible && shortableOption.codePointAt(0) === EQUAL_CODE) {
            hasShortValueSeparator = true
          } else {
            expanded.push(`${SHORT_OPTION_PREFIX}${shortableOption}`)
          }
        }
      }
      if (shortValue) {
        expanded.push(shortValue)
      }
      remainings.unshift(...expanded)
      groupCount = expanded.length
      continue
    }

    if (isLongOption(arg)) {
      // e.g. `--foo`
      const longOption = arg.slice(2)
      tokens.push({
        kind: 'option',
        name: longOption,
        rawName: arg,
        index,
        value: undefined,
        inlineValue: undefined
      })

      continue
    }

    if (isLongOptionAndValue(arg)) {
      // e.g. `--foo=bar`
      const equalIndex = arg.indexOf(EQUAL_CHAR)
      const longOption = arg.slice(2, equalIndex)
      const value = arg.slice(equalIndex + 1)
      tokens.push({
        kind: 'option',
        name: longOption,
        rawName: `${LONG_OPTION_PREFIX}${longOption}`,
        index,
        value,
        inlineValue: true
      })
      continue
    }

    tokens.push({
      kind: 'positional',
      index,
      value: arg
    })
  }

  return tokens
}

/**
 * Check if `arg` is a short option (e.g. `-f`).
 * @param arg the argument to check
 * @returns whether `arg` is a short option.
 */
export function isShortOption(arg: string): boolean {
  return (
    arg.length === 2 && arg.codePointAt(0) === HYPHEN_CODE && arg.codePointAt(1) !== HYPHEN_CODE
  )
}

/**
 * Check if `arg` is a short option group (e.g. `-abc`).
 * @param arg the argument to check
 * @returns whether `arg` is a short option group.
 */
function isShortOptionGroup(arg: string): boolean {
  if (arg.length <= 2) {
    return false
  }

  if (arg.codePointAt(0) !== HYPHEN_CODE) {
    return false
  }

  if (arg.codePointAt(1) === HYPHEN_CODE) {
    return false
  }

  return true
}

/**
 * Check if `arg` is a long option (e.g. `--foo`).
 * @param arg the argument to check
 * @returns whether `arg` is a long option.
 */
function isLongOption(arg: string) {
  return hasLongOptionPrefix(arg) && !arg.includes(EQUAL_CHAR, 3)
}

/**
 * Check if `arg` is a long option with value (e.g. `--foo=bar`).
 * @param arg the argument to check
 * @returns whether `arg` is a long option.
 */
function isLongOptionAndValue(arg: string) {
  return hasLongOptionPrefix(arg) && arg.includes(EQUAL_CHAR, 3)
}

/**
 * Check if `arg` is a long option prefix (e.g. `--`).
 * @param arg the argument to check
 * @returns whether `arg` is a long option prefix.
 */
export function hasLongOptionPrefix(arg: string): boolean {
  // @ts-ignore -- NOTE: `~` is a bitwise NOT operator
  return arg.length > 2 && ~arg.indexOf(LONG_OPTION_PREFIX)
}

/**
 * Check if a `value` is an option value.
 * @param value a value to check
 * @returns whether a `value` is an option value.
 */
function hasOptionValue(value: string | undefined): boolean {
  // eslint-disable-next-line unicorn/no-null
  return !(value == null) && value.codePointAt(0) !== HYPHEN_CODE
}
