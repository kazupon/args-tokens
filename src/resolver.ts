/**
 * Entry point of argument options resolver.
 * @module
 */

/**
 * @author kazuya kawaguchi (a.k.a. kazupon)
 * @license MIT
 */

import { hasLongOptionPrefix, isShortOption } from './parser.ts'

import type { ArgToken } from './parser.ts'

/**
 * An option schema for an argument.
 * This schema is similar to the schema of the `node:utils`.
 * difference is that:
 * - `multiple` property is not supported
 * - `required` property and `description` property are added
 * - `type` is not only 'string' and 'boolean', but also 'number' and 'enum' too.
 * - `default` property type, not support multiple types
 */
export interface ArgOptionSchema {
  /**
   * Type of argument.
   */
  type: 'string' | 'boolean' | 'number' | 'enum'
  /**
   * A single character alias for the option.
   */
  short?: string
  /**
   * A description of the argument.
   */
  description?: string
  /**
   * Whether the argument is required or not.
   */
  required?: true
  /**
   * The allowed values of the argument, and string only. This property is only used when the type is 'enum'.
   */
  choices?: string[]
  /**
   * The default value of the argument.
   * if the type is 'enum', the default value must be one of the allowed values.
   */
  default?: string | boolean | number
}

/**
 * An object that contains {@link ArgOptionSchema | options schema}.
 */
export interface ArgOptions {
  [option: string]: ArgOptionSchema
}

/**
 * An object that contains the values of the arguments.
 */
export type ArgValues<T> = T extends ArgOptions
  ? ResolveArgValues<
      T,
      {
        [Option in keyof T]: ExtractOptionValue<T[Option]>
      }
    >
  : {
      [option: string]: string | boolean | number | undefined
    }

/**
 * @internal
 */
export type ExtractOptionValue<O extends ArgOptionSchema> = O['type'] extends 'string'
  ? string
  : O['type'] extends 'boolean'
    ? boolean
    : O['type'] extends 'number'
      ? number
      : O['type'] extends 'enum'
        ? O['choices'] extends string[]
          ? O['choices'][number]
          : never
        : string | boolean | number

/**
 * @internal
 */
export type ResolveArgValues<O extends ArgOptions, V extends Record<keyof O, unknown>> = {
  -readonly [Option in keyof O]?: V[Option]
} & FilterArgs<O, V, 'default'> &
  FilterArgs<O, V, 'required'> extends infer P
  ? { [K in keyof P]: P[K] }
  : never

/**
 * @internal
 */
export type FilterArgs<
  O extends ArgOptions,
  V extends Record<keyof O, unknown>,
  K extends keyof ArgOptionSchema
> = {
  // eslint-disable-next-line @typescript-eslint/no-empty-object-type
  [Option in keyof O as O[Option][K] extends {} ? Option : never]: V[Option]
}

/**
 * An options for {@link resolveArgs | resolve arguments}.
 */
export interface ResolveArgsOptions {
  /**
   * Whether to group short options.
   * @default false
   * @see guideline 5 in https://pubs.opengroup.org/onlinepubs/9799919799/basedefs/V1_chap12.html
   */
  optionGrouping?: boolean
}

/**
 * Resolve command line arguments.
 * @param options - An options that contains {@link ArgOptionSchema | options schema}.
 * @param tokens - An array of {@link ArgToken | tokens}.
 * @param resolveArgsOptions - An options that contains {@link resolveArgsOptions | resolve arguments options}.
 * @returns An object that contains the values of the arguments, positional arguments, and {@link AggregateError | validation errors}.
 */
export function resolveArgs<T extends ArgOptions>(
  options: T,
  tokens: ArgToken[],
  { optionGrouping = false }: ResolveArgsOptions = {}
): { values: ArgValues<T>; positionals: string[]; error: AggregateError | undefined } {
  const positionals = [] as string[]

  const longOptionTokens: ArgToken[] = []
  const shortOptionTokens: ArgToken[] = []

  let currentLongOption: ArgToken | undefined
  let currentShortOption: ArgToken | undefined
  const expandableShortOptions: ArgToken[] = []

  function toShortValue(): string | undefined {
    if (expandableShortOptions.length === 0) {
      return undefined
    } else {
      const value = expandableShortOptions.map(token => token.name).join('')
      expandableShortOptions.length = 0
      return value
    }
  }

  function applyLongOptionValue(value: string | undefined = undefined): void {
    if (currentLongOption) {
      currentLongOption.value = value
      longOptionTokens.push({ ...currentLongOption })
      currentLongOption = undefined
    }
  }

  function applyShortOptionValue(value: string | undefined = undefined): void {
    if (currentShortOption) {
      currentShortOption.value = value || toShortValue()
      shortOptionTokens.push({ ...currentShortOption })
      currentShortOption = undefined
    }
  }

  /**
   * analyze phase to resolve value
   * separate tokens into positionals, long and short options, after that resolve values
   */

  const schemas = Object.values(options)

  // eslint-disable-next-line unicorn/no-for-loop
  for (let i = 0; i < tokens.length; i++) {
    const token = tokens[i]
    if (token.kind === 'positional') {
      if (currentShortOption) {
        const found = schemas.find(
          schema => schema.short === currentShortOption!.name && schema.type === 'boolean'
        )
        if (found) {
          positionals.push(token.value!)
        }
      } else if (currentLongOption) {
        const found = options[currentLongOption.name!]?.type === 'boolean'
        if (found) {
          positionals.push(token.value!)
        }
      } else {
        positionals.push(token.value!)
      }
      // check if previous option is not resolved
      applyLongOptionValue(token.value)
      applyShortOptionValue(token.value)
    } else if (token.kind === 'option') {
      if (token.rawName) {
        if (hasLongOptionPrefix(token.rawName)) {
          if (token.inlineValue) {
            longOptionTokens.push({ ...token })
          } else {
            currentLongOption = { ...token }
          }
          // check if previous short option is not resolved
          applyShortOptionValue()
        } else if (isShortOption(token.rawName)) {
          if (currentShortOption) {
            if (currentShortOption.index === token.index) {
              if (optionGrouping) {
                currentShortOption.value = token.value
                shortOptionTokens.push({ ...currentShortOption })
                currentShortOption = { ...token }
              } else {
                expandableShortOptions.push({ ...token })
              }
            } else {
              currentShortOption.value = toShortValue()
              shortOptionTokens.push({ ...currentShortOption })
              currentShortOption = { ...token }
            }
            // check if previous long option is not resolved
            applyLongOptionValue()
          } else {
            currentShortOption = { ...token }
            // check if previous long option is not resolved
            applyLongOptionValue()
          }
        }
      } else {
        // short option value
        if (currentShortOption && currentShortOption.index == token.index && token.inlineValue) {
          currentShortOption.value = token.value
          shortOptionTokens.push({ ...currentShortOption })
          currentShortOption = undefined
        }
        // check if previous long option is not resolved
        applyLongOptionValue()
      }
    } else {
      // check if previous option is not resolved
      applyLongOptionValue()
      applyShortOptionValue()
    }
  }

  /**
   * check if the last long or short option is not resolved
   */

  applyLongOptionValue()
  applyShortOptionValue()

  /**
   * resolve values
   */

  const values = Object.create(null) as ArgValues<T>
  const errors: Error[] = []

  for (const [option, schema] of Object.entries(options)) {
    if (schema.required) {
      const found =
        longOptionTokens.find(token => token.name === option) ||
        (schema.short && shortOptionTokens.find(token => token.name === schema.short))
      if (!found) {
        errors.push(createRequireError(option, schema))
        continue
      }
    }

    // eslint-disable-next-line unicorn/no-for-loop
    for (let i = 0; i < longOptionTokens.length; i++) {
      const token = longOptionTokens[i]
      // eslint-disable-next-line unicorn/no-null
      if (option === token.name && token.rawName != null && hasLongOptionPrefix(token.rawName)) {
        const invalid = validateRequire(token, option, schema)
        if (invalid) {
          errors.push(invalid)
          continue
        }

        if (schema.type === 'boolean') {
          // NOTE: re-set value to undefined, because long boolean type option is set on analyze phase
          token.value = undefined
        } else {
          const invalid = validateValue(token, option, schema)
          if (invalid) {
            errors.push(invalid)
            continue
          }
        }

        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        ;(values as any)[option] = resolveOptionValue(token, schema)
        continue
      }
    }

    // eslint-disable-next-line unicorn/no-for-loop
    for (let i = 0; i < shortOptionTokens.length; i++) {
      const token = shortOptionTokens[i]

      // eslint-disable-next-line unicorn/no-null
      if (schema.short === token.name && token.rawName != null && isShortOption(token.rawName)) {
        const invalid = validateRequire(token, option, schema)
        if (invalid) {
          errors.push(invalid)
          continue
        }

        if (schema.type === 'boolean') {
          // NOTE: re-set value to undefined, because short boolean type option is set on analyze phase
          token.value = undefined
        } else {
          const invalid = validateValue(token, option, schema)
          if (invalid) {
            errors.push(invalid)
            continue
          }
        }

        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        ;(values as any)[option] = resolveOptionValue(token, schema)
        continue
      }
    }

    // eslint-disable-next-line unicorn/no-null
    if (values[option] == null && schema.default != null) {
      // check if the default value is in values
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      ;(values as any)[option] = schema.default
    }
  }

  // eslint-disable-next-line unicorn/error-message
  return { values, positionals, error: errors.length > 0 ? new AggregateError(errors) : undefined }
}

function createRequireError(option: string, schema: ArgOptionSchema): OptionResolveError {
  return new OptionResolveError(
    `Option '--${option}' ${schema.short ? `or '-${schema.short}' ` : ''}is required`,
    option,
    'required',
    schema
  )
}

/**
 * An error type for {@link OptionResolveError}.
 */
export type OptionResolveErrorType = 'type' | 'required'

/**
 * An error that occurs when resolving options.
 * This error is thrown when the option is not valid.
 */
export class OptionResolveError extends Error {
  override name: string
  schema: ArgOptionSchema
  type: OptionResolveErrorType
  constructor(
    message: string,
    name: string,
    type: OptionResolveErrorType,
    schema: ArgOptionSchema
  ) {
    super(message)
    this.name = name
    this.type = type
    this.schema = schema
  }
}

function validateRequire(
  token: ArgToken,
  option: string,
  schema: ArgOptionSchema
): Error | undefined {
  if (schema.required && schema.type !== 'boolean' && !token.value) {
    return createRequireError(option, schema)
  }
}

function validateValue(
  token: ArgToken,
  option: string,
  schema: ArgOptionSchema
): Error | undefined {
  switch (schema.type) {
    case 'number': {
      if (!isNumeric(token.value!)) {
        return createTypeError(option, schema)
      }
      break
    }
    case 'string': {
      if (typeof token.value !== 'string') {
        return createTypeError(option, schema)
      }
      break
    }
    case 'enum': {
      if (schema.choices && !schema.choices.includes(token.value!)) {
        return new OptionResolveError(
          `Option '--${option}' ${schema.short ? `or '-${schema.short}' ` : ''}should be choiced from '${schema.type}' [${schema.choices.map(c => JSON.stringify(c)).join(', ')}] values`,
          option,
          'type',
          schema
        )
      }
      break
    }
  }
}

function isNumeric(str: string): boolean {
  // @ts-ignore
  // eslint-disable-next-line unicorn/prefer-number-properties
  return str.trim() !== '' && !isNaN(str)
}

function createTypeError(option: string, schema: ArgOptionSchema): TypeError {
  return new OptionResolveError(
    `Option '--${option}' ${schema.short ? `or '-${schema.short}' ` : ''}should be '${schema.type}'`,
    option,
    'type',
    schema
  )
}

function resolveOptionValue(
  token: ArgToken,
  schema: ArgOptionSchema
): string | boolean | number | undefined {
  if (token.value) {
    return schema.type === 'number' ? +token.value : token.value
  }

  if (schema.type === 'boolean') {
    return true
  }

  return schema.type === 'number' ? +(schema.default || '') : schema.default
}
