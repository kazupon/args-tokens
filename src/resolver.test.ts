import { describe, expect, test } from 'vitest'
import { parseArgs } from './parser.ts'
import { OptionResolveError, resolveArgs } from './resolver.ts'

import type { ArgOptions } from './resolver.ts'

const options = {
  help: {
    type: 'boolean',
    short: 'h'
  },
  version: {
    type: 'boolean',
    short: 'v'
  },
  port: {
    type: 'number',
    short: 'p',
    default: 8080
  },
  mode: {
    type: 'string',
    short: 'm'
  },
  host: {
    type: 'string',
    short: 'o',
    required: true
  }
} as const satisfies ArgOptions

describe('resolveArgs', () => {
  test('basic', () => {
    const args = ['dev', '--port=9131', '--host=example.com', '--help']
    const tokens = parseArgs(args)
    const { values, positionals, error } = resolveArgs(options, tokens)
    expect(values).toEqual({
      port: 9131,
      host: 'example.com',
      help: true
    })
    expect(positionals).toEqual(['dev'])
    expect(error).toBeUndefined()
  })

  test('missing required option', () => {
    const args = ['dev']
    const tokens = parseArgs(args)
    const { error } = resolveArgs(options, tokens)
    expect(error?.errors.length).toBe(1)
    expect((error?.errors[0] as Error).message).toEqual("Option '--host' or '-o' is required")
    expect((error?.errors[0] as OptionResolveError).name).toEqual('host')
    expect((error?.errors[0] as OptionResolveError).type).toEqual('required')
    expect((error?.errors[0] as OptionResolveError).schema.type).toEqual('string')
  })

  test('missing defaultable option', () => {
    const args = ['dev', '--host=example.com']
    const tokens = parseArgs(args)
    const { values, positionals } = resolveArgs(options, tokens)
    expect(values).toEqual({
      port: 8080,
      host: 'example.com'
    })
    expect(positionals).toEqual(['dev'])
  })

  test('invalid value', () => {
    const args = ['dev', '--port=foo', '--host=example.com']
    const tokens = parseArgs(args)
    const { error } = resolveArgs(options, tokens)
    expect(error?.errors.length).toBe(1)
    expect((error?.errors[0] as OptionResolveError).message).toEqual(
      "Option '--port' or '-p' should be 'number'"
    )
    expect((error?.errors[0] as OptionResolveError).name).toEqual('port')
    expect((error?.errors[0] as OptionResolveError).type).toEqual('type')
    expect((error?.errors[0] as OptionResolveError).schema.type).toEqual('number')
  })

  test('multiple errors', () => {
    const args = ['dev', '--port=foo']
    const tokens = parseArgs(args)
    const { error } = resolveArgs(options, tokens)
    expect(error?.errors.length).toBe(2)
    expect((error?.errors[0] as Error).message).toEqual(
      "Option '--port' or '-p' should be 'number'"
    )
    expect((error?.errors[0] as OptionResolveError).name).toEqual('port')
    expect((error?.errors[0] as OptionResolveError).type).toEqual('type')
    expect((error?.errors[1] as OptionResolveError).message).toEqual(
      "Option '--host' or '-o' is required"
    )
    expect((error?.errors[1] as OptionResolveError).name).toEqual('host')
    expect((error?.errors[1] as OptionResolveError).type).toEqual('required')
    expect((error?.errors[1] as OptionResolveError).schema.type).toEqual('string')
  })

  test('missing positionals', () => {
    const args = ['--port=9131', '--host=example.com']
    const tokens = parseArgs(args)
    const { values, positionals } = resolveArgs(options, tokens)
    expect(values).toEqual({
      port: 9131,
      host: 'example.com'
    })
    expect(positionals).toEqual([])
  })

  test('positionals', () => {
    const args = ['dev', '--host=example.com', 'foo', 'bar']
    const tokens = parseArgs(args)
    const { positionals, error } = resolveArgs(options, tokens)
    expect(positionals).toEqual(['dev', 'foo', 'bar'])
    expect(error).toBeUndefined()
  })

  test('long options value captured from positionals', () => {
    const args = ['dev', '--port', '9131', '--host', 'example.com', 'bar']
    const tokens = parseArgs(args)
    const { values, positionals } = resolveArgs(options, tokens)
    expect(values).toEqual({
      port: 9131,
      host: 'example.com'
    })
    expect(positionals).toEqual(['dev', 'bar'])
  })

  test('short options value specified with equals', () => {
    const args = ['dev', '-p=9131', '-o=example.com', '-h']
    const tokens = parseArgs(args)
    const { values, positionals, error } = resolveArgs(options, tokens)
    expect(values).toEqual({
      port: 9131,
      host: 'example.com',
      help: true
    })
    expect(positionals).toEqual(['dev'])
    expect(error).toBeUndefined()
  })

  test('short options value specified with concatenation', () => {
    const args = ['dev', '-p9131', '-oexample.com']
    const tokens = parseArgs(args)
    const { values, positionals, error } = resolveArgs(options, tokens)
    expect(values).toEqual({
      port: 9131,
      host: 'example.com'
    })
    expect(positionals).toEqual(['dev'])
    expect(error).toBeUndefined()
  })

  test('short options value captured from positionals', () => {
    const args = ['dev', '-p', '9131', '-o', 'example.com', 'bar']
    const tokens = parseArgs(args)
    const { values, positionals, error } = resolveArgs(options, tokens)
    expect(values).toEqual({
      port: 9131,
      host: 'example.com'
    })
    expect(positionals).toEqual(['dev', 'bar'])
    expect(error).toBeUndefined()
  })

  test('complex options', () => {
    const args = [
      'dev',
      '-p9131',
      '--host',
      'example.com',
      'foo',
      '-m=production',
      '-h',
      '--version',
      'bar',
      'baz'
    ]
    const tokens = parseArgs(args)
    const { values, positionals, error } = resolveArgs(options, tokens)
    expect(values).toEqual({
      port: 9131,
      host: 'example.com',
      mode: 'production',
      help: true,
      version: true
    })
    expect(positionals).toEqual(['dev', 'foo', 'bar', 'baz'])
    expect(error).toBeUndefined()
  })

  test('sanitize options', () => {
    const args = ['dev', '--__proto__', '{ "polluted": 1 }']
    const tokens = parseArgs(args)
    const { values, positionals, error } = resolveArgs(
      {
        __proto__: {
          type: 'string'
        },
        foo: {
          type: 'string',
          default: 'foo'
        }
      },
      tokens
    )
    expect(values).toEqual({
      foo: 'foo'
    })
    expect(positionals).toEqual(['dev'])
    expect(error).toBeUndefined()
  })
})

describe('option group', () => {
  test('basic', () => {
    const args = ['dev', '-dsV']
    const tokens = parseArgs(args)
    const { values, positionals } = resolveArgs(
      {
        debug: {
          type: 'boolean',
          short: 'd'
        },
        silent: {
          type: 'boolean',
          short: 's'
        },
        verbose: {
          type: 'boolean',
          short: 'V'
        }
      },
      tokens,
      { optionGrouping: true }
    )
    expect(positionals).toEqual(['dev'])
    expect(values).toEqual({
      debug: true,
      silent: true,
      verbose: true
    })
  })

  test('mix option grouping and long option', () => {
    const args = ['dev', '-ds', '--host', 'example.com', '-Vm', 'foo', 'bar']
    const tokens = parseArgs(args)
    const { values, positionals } = resolveArgs(
      {
        debug: {
          type: 'boolean',
          short: 'd'
        },
        silent: {
          type: 'boolean',
          short: 's'
        },
        verbose: {
          type: 'boolean',
          short: 'V'
        },
        host: {
          type: 'string',
          short: 'o'
        },
        minify: {
          type: 'boolean',
          short: 'm'
        }
      },
      tokens,
      { optionGrouping: true }
    )
    expect(positionals).toEqual(['dev', 'foo', 'bar'])
    expect(values).toEqual({
      debug: true,
      silent: true,
      verbose: true,
      minify: true,
      host: 'example.com'
    })
  })
})

describe('enum option', () => {
  test('basic', () => {
    const args = ['dev', '--log=debug']
    const tokens = parseArgs(args)
    const { values, positionals } = resolveArgs(
      {
        log: {
          type: 'enum',
          short: 'l',
          choices: ['debug', 'info', 'warn', 'error']
        }
      },
      tokens
    )
    expect(values).toEqual({
      log: 'debug'
    })
    expect(positionals).toEqual(['dev'])
  })

  test('invalid value', () => {
    const args = ['dev', '--log=foo']
    const tokens = parseArgs(args)
    const { error } = resolveArgs(
      {
        log: {
          type: 'enum',
          short: 'l',
          choices: ['debug', 'info', 'warn', 'error']
        }
      },
      tokens
    )
    expect(error?.errors.length).toBe(1)
    expect((error?.errors[0] as OptionResolveError).message).toEqual(
      `Option '--log' or '-l' should be choiced from 'enum' ["debug", "info", "warn", "error"] values`
    )
    expect((error?.errors[0] as OptionResolveError).name).toEqual('log')
    expect((error?.errors[0] as OptionResolveError).type).toEqual('type')
    expect((error?.errors[0] as OptionResolveError).schema.type).toEqual('enum')
  })

  test('required', () => {
    const args = ['dev']
    const tokens = parseArgs(args)
    const { error } = resolveArgs(
      {
        log: {
          type: 'enum',
          short: 'l',
          choices: ['debug', 'info', 'warn', 'error'],
          required: true
        }
      },
      tokens
    )
    expect(error?.errors.length).toBe(1)
    expect((error?.errors[0] as Error).message).toEqual("Option '--log' or '-l' is required")
    expect((error?.errors[0] as OptionResolveError).name).toEqual('log')
    expect((error?.errors[0] as OptionResolveError).type).toEqual('required')
    expect((error?.errors[0] as OptionResolveError).schema.type).toEqual('enum')
  })

  test('missing', () => {
    const args = ['dev', '--', 'bar']
    const tokens = parseArgs(args)
    const { error, values, positionals } = resolveArgs(
      {
        log: {
          type: 'enum',
          short: 'l',
          choices: ['debug', 'info', 'warn', 'error']
        }
      },
      tokens
    )
    expect(error).toBeUndefined()
    expect(values).toEqual({})
    expect(positionals).toEqual(['dev', 'bar'])
  })

  test('default', () => {
    const args = ['dev']
    const tokens = parseArgs(args)
    const { values, positionals } = resolveArgs(
      {
        log: {
          type: 'enum',
          short: 'l',
          choices: ['debug', 'info', 'warn', 'error'],
          default: 'info'
        }
      },
      tokens
    )
    expect(values).toEqual({
      log: 'info'
    })
    expect(positionals).toEqual(['dev'])
  })

  test('invalid default', () => {
    const args = ['dev', '--log']
    const tokens = parseArgs(args)
    const { error } = resolveArgs(
      {
        log: {
          type: 'enum',
          short: 'l',
          choices: ['debug', 'info', 'warn', 'error'],
          default: 'foo'
        }
      },
      tokens
    )
    expect(error?.errors.length).toBe(1)
    expect((error?.errors[0] as OptionResolveError).message).toEqual(
      `Option '--log' or '-l' should be choiced from 'enum' ["debug", "info", "warn", "error"] values`
    )
    expect((error?.errors[0] as OptionResolveError).name).toEqual('log')
    expect((error?.errors[0] as OptionResolveError).type).toEqual('type')
    expect((error?.errors[0] as OptionResolveError).schema.type).toEqual('enum')
  })
})
