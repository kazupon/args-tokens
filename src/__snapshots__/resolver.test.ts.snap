// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`resolveArgs > invalid value 1`] = `[TypeError: Option '--port' or '-p' should be 'number']`;

exports[`resolveArgs > missing required option 1`] = `[Error: Option '--host' or '-o' is required]`;
